src/solvers/refinement/string_constraint_generator_constants.cpp

```diff
@@ -81,7 +114,5 @@ exprt string_constraint_generatort::add_axioms_from_literal(
   const function_application_exprt::argumentst &args=f.arguments();
   PRECONDITION(args.size() == 3); // Bad args to string literal?
   const array_string_exprt res = char_array_of_pointer(args[1], args[0]);
-  const exprt &arg = args[2];
-  irep_idt sval=to_constant_expr(arg).get_value();
-  return add_axioms_for_constant(res, sval);
+  return add_axioms_for_cprover_string(res, args[2], true_exprt());
```

Comment:
It would be good to see a test for whatever case is now supported. I can't say the changes in this file mean much to me

Reply:
There will be tests in other repositories: https://github.com/diffblue/test-gen/pull/1327
But it is difficule to test on CBMC because we don't have a model for Class and the main affected method is Object.getClass.

An example that should work afterwards is the following:
```java
public class testMaybeNull1
{
    public Object x;
    public String check()
    {
        return x.getClass().getName();
    }
}
```

The reason is that getClass converts a cprover_string to a java string (the `@class_identifier`), but since the object could be null, the argument passed to `cprover_string_literal` is more complicated than a constant.

Reply:
Would it not be sufficent to have a:

```java
public class simpleClaz {
  String getStr() { return "hello"; }
}
/////////////
public class testMaybeNull1
{
    public simpleClaz x;
    public String check()
    {
        return x.getStr();
    }
}
```
Since x may or may not be null isn't it the same?

Reply:
No this example is already handled correctly. It is the `@class_identifier` field that is a problem, and as far as I now it can only be access via `instanceof` and `getClass`. And I think `getClass` is the only function which generate a `cprover_string_literal` with something else than a constant.

Reply:
> But it is difficule to test on CBMC because we don't have a model for Class and the main affected method is Object.getClass.

Getting 'java core models' into the cbmc repo would help. @cesaro has started some work on that.